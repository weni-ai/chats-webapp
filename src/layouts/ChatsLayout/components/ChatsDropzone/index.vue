<template>
  <div
    ref="dropzone"
    class="dropzone"
    :class="{
      dropzone: true,
      dragging: show && isDragging,
    }"
    @dragenter.stop.prevent="dragenter"
    @dragover.stop.prevent="dragover"
    @dragleave.stop.prevent="dragleave"
    @drop.stop.prevent="drop"
  >
    <div v-if="show && isDragging" class="dropzone__description">
      <unnnic-icon-svg
        class="unnnic-upload-area__dropzone__icon"
        icon="upload-bottom-1"
        scheme="brand-weni"
        size="xl"
      />
      <h1>{{ $t('dropzone.description') }}</h1>
    </div>

    <slot />
  </div>
</template>

<script>
export default {
  name: 'ChatsDropzone',

  props: {
    show: {
      type: Boolean,
      default: true,
    },
  },

  data() {
    return {
      files: [],
      hasError: false,
      isDragging: false,
      dragEnterCounter: 0, // to handle dragenter/dragleave on child elements
    };
  },

  methods: {
    openFileUploader(files) {
      this.$emit('open-file-uploader', files);
    },

    dragenter() {
      this.dragEnterCounter += 1;
      this.isDragging = true;
    },

    dragover() {
      this.isDragging = true;
    },

    dragleave() {
      this.dragEnterCounter -= 1;

      if (this.dragEnterCounter === 0) {
        this.isDragging = false;
      }
    },

    drop(event) {
      const { files } = event.dataTransfer;
      if (files.length > 0) this.openFileUploader(files);

      this.dragEnterCounter = 0;
      this.isDragging = false;
    },
  },
};
</script>

<style lang="scss" scoped>
@function borderDashed($color) {
  $colorString: unquote('' + $color);
  $cleanColor: str-slice($colorString, 2);
  @return url("data:image/svg+xml,%3csvg width='100%25' height='100%25' xmlns='http://www.w3.org/2000/svg'%3e%3crect width='100%25' height='100%25' fill='none' rx='16' ry='16' stroke='%23#{$cleanColor}' stroke-width='4' stroke-dasharray='4%2c 12' stroke-dashoffset='9' stroke-linecap='square'/%3e%3c/svg%3e");
}

.dropzone {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: $unnnic-spacing-nano;

  height: 100%;

  overflow: hidden;

  padding-left: $unnnic-spacing-inline-sm;

  &.dragging {
    position: relative;

    border-radius: $unnnic-border-radius-lg;

    margin: $unnnic-spacing-inset-sm;
    margin-bottom: 0;

    padding-left: 0;

    background-color: $unnnic-color-neutral-white;

    &::before,
    &::after {
      content: '';
    }

    &::before {
      background-color: $unnnic-color-background-snow;
    }

    &::after {
      border-radius: $unnnic-border-radius-lg;
      // Dashed border with increased dashes spacing and color neutral clean
      background-image: borderDashed($unnnic-color-brand-weni);
    }
  }

  &__description,
  &.dragging::before,
  &.dragging::after {
    z-index: 1000;
    position: absolute;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
  }

  &__description {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    gap: $unnnic-spacing-stack-xs;

    h1 {
      font-size: $unnnic-font-size-title-sm;
      font-weight: $unnnic-font-weight-bold;
      color: $unnnic-color-brand-weni;
    }
  }
}
</style>
