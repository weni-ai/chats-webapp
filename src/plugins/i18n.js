import Vue from 'vue';
import VueI18n from 'vue-i18n';

// eslint-disable-next-line camelcase
import pt_br from '@/locales/pt_br.json';
import en from '@/locales/en.json';
import es from '@/locales/es.json';

import moment from 'moment';

import('moment/locale/pt-br');
import('moment/locale/es');

moment.locale('pt-br');

Vue.use(VueI18n);

const messages = {
  'pt-br': pt_br,
  en,
  es,
};

const i18n = new VueI18n({
  locale: 'pt-br',
  fallbackLocale: 'pt-br',
  messages,
  dateTimeFormats: {
    'pt-br': {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    en: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    es: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
  },
});

export default i18n;
