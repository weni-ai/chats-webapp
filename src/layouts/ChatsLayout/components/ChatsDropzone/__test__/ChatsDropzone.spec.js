import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';
import { mount, config } from '@vue/test-utils';
import ChatsDropzone from '@/layouts/ChatsLayout/components/ChatsDropzone/index.vue';
import i18n from '@/plugins/i18n';

beforeAll(() => {
  config.global.plugins = config.global.plugins.filter(
    (plugin) => plugin !== i18n,
  );
});

afterAll(() => {
  if (!config.global.plugins.includes(i18n)) {
    config.global.plugins.push(i18n);
  }
});

describe('ChatsDropzone', () => {
  const createWrapper = (props = {}, slots = {}) =>
    mount(ChatsDropzone, {
      props: { show: true, ...props },
      slots,
      global: {
        mocks: { $t: (key) => key },
        stubs: {
          UnnnicIconSvg: {
            template: '<span data-testid="icon-svg"></span>',
            props: ['icon', 'scheme', 'size'],
          },
        },
      },
    });

  const mockEvent = (files = []) => ({ dataTransfer: { files } });
  const mockFile = (name = 'test.txt') => ({
    name,
    type: 'text/plain',
    size: 1024,
  });

  describe('Props and Initial State', () => {
    it('should handle props and initialize correctly', () => {
      const wrapper = createWrapper();
      expect(wrapper.props('show')).toBe(true);
      expect(wrapper.vm.files).toEqual([]);
      expect(wrapper.vm.isDragging).toBe(false);
      expect(wrapper.vm.dragEnterCounter).toBe(0);

      const wrapperHidden = createWrapper({ show: false });
      expect(wrapperHidden.props('show')).toBe(false);
    });
  });

  describe('Rendering', () => {
    it('should render container and slot content', () => {
      const wrapper = createWrapper(
        {},
        { default: '<div data-testid="slot">Content</div>' },
      );
      const container = wrapper.find('[data-testid="dropzone-container"]');

      expect(container.exists()).toBe(true);
      expect(container.classes()).toContain('dropzone');
      expect(wrapper.find('[data-testid="slot"]').exists()).toBe(true);
    });

    it('should conditionally show drag description and apply classes', async () => {
      const wrapper = createWrapper();

      // Initially hidden
      expect(
        wrapper.find('[data-testid="dropzone-description"]').exists(),
      ).toBe(false);

      // Show when dragging
      await wrapper.setData({ isDragging: true });
      const container = wrapper.find('[data-testid="dropzone-container"]');
      const description = wrapper.find('[data-testid="dropzone-description"]');

      expect(description.exists()).toBe(true);
      expect(wrapper.find('[data-testid="dropzone-icon"]').exists()).toBe(true);
      expect(wrapper.find('[data-testid="dropzone-title"]').text()).toBe(
        'dropzone.description',
      );
      expect(container.classes()).toContain('dragging');

      // Hidden when show=false
      const wrapperHidden = createWrapper({ show: false });
      await wrapperHidden.setData({ isDragging: true });
      expect(
        wrapperHidden.find('[data-testid="dropzone-description"]').exists(),
      ).toBe(false);
      expect(
        wrapperHidden.find('[data-testid="dropzone-container"]').classes(),
      ).not.toContain('dragging');
    });
  });

  describe('Drag and Drop Functionality', () => {
    it('should handle complete drag sequence with counter logic', async () => {
      const wrapper = createWrapper();
      const container = wrapper.find('[data-testid="dropzone-container"]');

      // dragenter - increment counter, set dragging
      await container.trigger('dragenter');
      expect(wrapper.vm.dragEnterCounter).toBe(1);
      expect(wrapper.vm.isDragging).toBe(true);

      // Multiple dragenter events
      await container.trigger('dragenter');
      expect(wrapper.vm.dragEnterCounter).toBe(2);

      // dragover - maintain dragging state
      await wrapper.setData({ isDragging: false });
      await container.trigger('dragover');
      expect(wrapper.vm.isDragging).toBe(true);

      // dragleave - decrement counter, stop dragging when counter = 0
      await wrapper.setData({ dragEnterCounter: 2 });
      await container.trigger('dragleave');
      expect(wrapper.vm.dragEnterCounter).toBe(1);
      expect(wrapper.vm.isDragging).toBe(true);

      await container.trigger('dragleave');
      expect(wrapper.vm.dragEnterCounter).toBe(0);
      expect(wrapper.vm.isDragging).toBe(false);
    });

    it('should handle file drop with proper emission and state reset', async () => {
      const wrapper = createWrapper();
      const container = wrapper.find('[data-testid="dropzone-container"]');
      const files = [mockFile(), mockFile('test2.txt')];

      // Drop with files - should emit and reset
      await wrapper.setData({ dragEnterCounter: 2, isDragging: true });
      await container.trigger('drop', mockEvent(files));

      expect(wrapper.emitted('open-file-uploader')).toBeTruthy();
      expect(wrapper.emitted('open-file-uploader')[0][0]).toBe(files);
      expect(wrapper.vm.dragEnterCounter).toBe(0);
      expect(wrapper.vm.isDragging).toBe(false);

      // Drop without files - should reset state but not emit
      await wrapper.setData({ dragEnterCounter: 1, isDragging: true });
      await container.trigger('drop', mockEvent([]));

      expect(wrapper.emitted('open-file-uploader')).toHaveLength(1); // Still only 1 emission
      expect(wrapper.vm.dragEnterCounter).toBe(0);
      expect(wrapper.vm.isDragging).toBe(false);
    });

    it('should handle direct method calls', () => {
      const wrapper = createWrapper();
      const files = [mockFile()];

      // Test openFileUploader method
      wrapper.vm.openFileUploader(files);
      expect(wrapper.emitted('open-file-uploader')).toBeTruthy();
      expect(wrapper.emitted('open-file-uploader')[0][0]).toBe(files);

      // Test drop method with and without files
      const dropSpy = vi.spyOn(wrapper.vm, 'openFileUploader');
      wrapper.vm.drop(mockEvent(files));
      expect(dropSpy).toHaveBeenCalledWith(files);

      wrapper.vm.drop(mockEvent([]));
      expect(dropSpy).toHaveBeenCalledTimes(1); // No additional call for empty files
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot', async () => {
      const wrapper = createWrapper();
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
