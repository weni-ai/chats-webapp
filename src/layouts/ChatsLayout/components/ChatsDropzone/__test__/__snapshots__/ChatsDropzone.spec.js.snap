// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatsDropzone > Snapshot > should match snapshot 1`] = `
<div
  class="dropzone dropzone"
  data-testid="dropzone-container"
  data-v-b64070c4=""
>
  <!--v-if-->
  
  
</div>
`;
